test/Functional.hs

```diff
@@ -1,226 +1,56 @@
--- ---------------------------------------------------------------------
-
-plugins :: IdePlugins
-plugins = pluginDescToIdePlugins
-  [("applyrefact", applyRefactDescriptor)
-  ,("eg2"        , example2Descriptor)
-  ,("ghcmod"     , ghcmodDescriptor)
-  ,("hare"       , hareDescriptor)
-  ,("base"       , baseDescriptor)
-  ]
-
-startServer :: IO (TChan (PluginRequest IO),TChan LogVal)
-startServer = do
-  cin      <- atomically newTChan
-  logChan  <- atomically newTChan
-
-  cancelTVar      <- atomically $ newTVar S.empty
-  wipTVar         <- atomically $ newTVar S.empty
-  versionTVar     <- atomically $ newTVar Map.empty
-  let dispatcherEnv = DispatcherEnv
-        { cancelReqsTVar     = cancelTVar
-        , wipReqsTVar        = wipTVar
-        , docVersionTVar     = versionTVar
-        }
-
-  void $ forkIO $ dispatcherP cin plugins testOptions dispatcherEnv
-                    (\lid errCode e -> logToChan logChan ("received an error", Left (lid, errCode, e)))
-                    (\g x -> g x)
-  return (cin,logChan)
-
--- ---------------------------------------------------------------------
-
-type LogVal = (String, Either (LspId, ErrorCode, T.Text) DynamicJSON)
-
-logToChan :: TChan LogVal -> LogVal -> IO ()
-logToChan c t = atomically $ writeTChan c t
-
--- ---------------------------------------------------------------------
-
-dispatchGhcRequest :: ToJSON a
-                   => TrackingNumber -> String -> Int
-                   -> TChan (PluginRequest IO) -> TChan LogVal
-                   -> PluginId -> CommandName -> a -> IO ()
-dispatchGhcRequest tn ctx n cin lc plugin com arg = do
-  let
-    logger :: RequestCallback IO DynamicJSON
-    logger x = logToChan lc (ctx, Right x)
-
-  let req = GReq tn Nothing Nothing (Just (IdInt n)) logger $
-        runPluginCommand plugin com (toJSON arg)
-  atomically $ writeTChan cin req
-
-dispatchIdeRequest :: (Typeable a, ToJSON a)
-                   => TrackingNumber -> String -> TChan (PluginRequest IO)
-                   -> TChan LogVal -> LspId -> IdeM (IdeResponse a) -> IO () 
-dispatchIdeRequest tn ctx cin lc lid f = do
-  let
-    logger :: (Typeable a, ToJSON a) => RequestCallback IO a
-    logger x = logToChan lc (ctx, Right (toDynJSON x))
-
-  let req = IReq tn lid logger f
-  atomically $ writeTChan cin req
-
--- ---------------------------------------------------------------------
+import Control.Monad.IO.Class
+import Control.Lens hiding (List)
+import Data.Aeson
+import qualified Data.HashMap.Strict as H
+import Language.Haskell.LSP.Test
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types as LSP (error, id)
+import Test.Hspec
+import System.Directory
+import System.FilePath
+import FunctionalDispatch
 
 main :: IO ()
 main = do
-  setupStackFiles
-  let logfile = "./test-functional.log"
-  exists <- doesFileExist logfile
-  when exists $ removeFile logfile
-  withFileLogging logfile $ cdAndDo "./test/testdata"  $ hspec spec
+  hspec spec
```

Comment:
`dispatcherSpec` calls `setupStackFiles`, which makes sure the test projects are compatible with the version of hie we are running.

It is probably best to keep all that logic here, as it was before.